### ISSUE TITLE
Type resolution for union between TypedDict and compatible dict

### ISSUE THREAD

**Issue by sayandipdutta** [microsoft/pyright#9833]
Here’s a piece of code that passes with mypy, but fails with pyright:
```python
from typing import TypedDict

class Foo(TypedDict):
    bar: int
    baz: list[int]

x: Foo = {"bar": 1, "baz": [1, 2, 3]}
y: Foo = x | {"baz": [1, 2, 3]}  # pyright error
```

pyright gives the following [error](https://pyright-play.net/?pythonVersion=3.14&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6tQBiYYAFEQjKUaASgBctKFKgAjZiDGYUdabOYAvRYySsYAbVQwAuvVoAPRdzBQAvFADeAIjkhHigIwAaKM41uoel5QAEzeAMxGAL60cJY8tlBmUAA%2BDr7q-oHeoVAR0UA):
```shell
error: Type "dict[str, object]" is not assignable to declared type "Foo"
    "dict[str, object]" is not assignable to "Foo" (reportAssignmentType)
1 error, 0 warnings, 0 informations
```

I imagine this should be valid as long as the TypedDict is the superset of the other operand.

[previous discussion](https://discuss.python.org/t/type-resolution-for-union-between-typeddict-and-dict/79663)

**Comment by erictraut** [comment]
The behavior of the `|` operator for a TypedDict is defined in the typeshed `typing.pyi` stub. Search for `_TypedDict`. 

For reference, here's the actual definition of the `__or__` method:

```python
    @overload
    def __or__(self, value: typing_extensions.Self, /) -> typing_extensions.Self: ...
    @overload
    def __or__(self, value: dict[str, Any], /) -> dict[str, object]: ...
```

Based on this definition, the expression that you're assigning to `y` in the example above evaluates to `dict[str, object]`, and that's not assignable to `Foo`.

Adding the feature you've requested would require hard-coding the behavior in the type checker and overriding the definition in the typeshed stub. As a general rule, I don't like to override typeshed definitions. I've broken this rule a few times, and I inevitable come to regret the decision because it leads to inconsistent and surprising behaviors when the typeshed stubs are invariably updated.

I suspect that mypy maintainers have chosen to hard-code the behavior rather than following the typeshed definitions.

Correctly annotating the `__or__` method in this case requires a facility that doesn't yet existing in the type system — a "Partial" operator. This has been discussed recently in the typing forums, and I'm hopeful that it will eventually be added to the type system. Once it is, the typeshed stubs can be updated to use it, and this issue will no longer exist.

I'm going to reject this enhancement request for now. I'm willing to revisit this in the future if there's sufficient signal from other pyright users (e.g. in the form of upvotes on this issue). I'll close the issue to make it clear that there are currently no plans to implement this.
